### PR TITLE
Add serializer mode for test262 execution

### PR DESCRIPTION
**Pull request by calebmer**
I want to better understand the bugs we have in abstract evaluation. One of my ideas to do this is to replace all the literals in test262 tests with abstract values and see what our test coverage is. The first step to do this is to serialize the test262 sources after running them in Prepack and checking if their generated JavaScript runs correctly. This PR does that by adding a `--serializer` flag.

With my methodology, all test262 harnesses are serialized every time since they are in the global scope. This unfortunately seems to be unavoidable since doing otherwise would change program semantics and break the tests.

Running `time yarn test-test262` takes about 2 minutes and has a 98% pass rate. Running `time yarn test-test262 --serializer` takes about 5 minutes and has a 95% pass rate. [Here’s a diff of the two  results.](https://gist.github.com/calebmer/1c9fe396b63ba055458c599c2be18a58)

[Here is a selection of some of the bugs](https://gist.github.com/calebmer/1ac381096a4aa7be1fc7dc2163276ab4) in the serializer caught by running test262 with the Prepack serializer. I might open issues for these, but they can be a bit pedantic. I might fix some of them depending on how important they are to the React code we want to compile.

Notably we have:

- Lots of invariants being triggered. I particularly saw [a lot of this invariant](https://github.com/facebook/prepack/blob/7d355ef4c5f4d939742814714a78ef9b1279f9b4/src/serializer/ResidualHeapVisitor.js#L531). (An example of this is [`07.md`](https://gist.github.com/calebmer/1ac381096a4aa7be1fc7dc2163276ab4#file-07-md).)
- Lots of values that are visited, but not serialized. (An example of this is [`01.md`](https://gist.github.com/calebmer/1ac381096a4aa7be1fc7dc2163276ab4#file-01-md).)
- Some incorrect outputs only caught by executing code. Could not be caught by static analysis. Particularly around class serialization.

I’ll build off this PR to get coverage for the abstract evaluator next, but I found these results interesting on their own so decided to do this in two PRs.

### CODE REVIEW

**Comment by NTillmann** [comment]
Regarding "all native function values should be intrinsics": There was a discussion on that here #1285 --- maybe that issue got closed prematurely.

**Comment by NTillmann** [comment]
> Lots of values that are visited, but not serialized.

That's not good. However, your examples seem to all involve classes, which wouldn't surprise me as that's an undertested feature (it's not ES5), and comes with an open issue mentioning some of the things that need to happen, possibly unrelated to the visiting/serialization bug you found: #1317.

**Comment by NTillmann** [comment]
> Some incorrect outputs only caught by executing code. Could not be caught by static analysis. Particularly around class serialization.

Hm, yes, classes... At the very least, Prepack should give up when encountering not actually supported class features. Great to see that we now have a way to detect them!

**Comment by calebmer** [comment]
“Lots” might be a bit of an exaggeration as you note, @NTillmann. More like “lots” relative to 5% of tests that failed. A 95% pass rate is pretty good.

**Comment by NTillmann** [comment]
What the pass rate when you exclude tests that involve classes?

File: scripts/test262-runner.js [hunk 1]
```diff
@@ -1088,6 +1129,82 @@ function runTest(
...
+      },
+    });
+  } catch (error) {
+    if (error.message === "Timed out") return new TestResult(false, strict, error);
+    if (error.message.includes("Syntax error")) return null;
+    return new TestResult(false, strict, new Error(`Prepack error:\n${error.stack}`));
+  }
+
+  const context = vm.createContext({
+    // NOTE: Workaround since Prepack serializes code that expects a global `TypedArray` class which does not exist per
```
Comment (NTillmann) [inline scripts/test262-runner.js]: Maybe it exists in JSC? Anyway, sounds like a bug that deserves an issue number and a TODO...

File: scripts/test262-runner.js [hunk 2]
```diff
@@ -1088,6 +1129,82 @@ function runTest(
...
+    result = prepackSources(sources, {
+      serialize: true,
+      timeout: timeout * 1000,
+      errorHandler: diag => {
+        if (diag.severity === "Information") return "Recover";
+        if (diag.severity !== "Warning") return "Fail";
+        return "Recover";
+      },
+      onParse: () => {
+        // TODO(calebmer): Turn all literals into abstract values
```
Comment (NTillmann) [inline scripts/test262-runner.js]: I guess behind some other option(s) that would allow to different levels of fuzzing?

Comment (calebmer) [reply scripts/test262-runner.js]: 👍

Comment (calebmer) [reply scripts/test262-runner.js]: Yep

File: scripts/test262-runner.js (hunk 1 above)
Comment (calebmer) [reply scripts/test262-runner.js]: #2292 👍

**Comment by calebmer** [comment]
> What's the change in pass rate when you exclude tests that involve classes?

95% still. I added a basic check to exclude source files that include `class`. ES6 passes rise from 91% to 93% though.

```
=== RESULTS ===
Passes: 16238 / 16949 (95%)
ES5 passes: 11501 / 11873 (96%)
ES6 passes: 4737 / 5076 (93%)
Skipped: 14206
Timeouts: 0
```

---

I also opened some issues for problems I thought were actionable and not class related. (Since class serialization is known to have issues.)

- #2292
- #2293
- #2294

None seem super related to React code, but #2293 might be. Haven’t audited all the `NativeFunctionValue`s this effects.

**Pull request closed without merging**